From bc6e1a0b1a4fff8688229c26847ca28063f128f0 Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Wed, 5 Jul 2023 15:24:23 +0200
Subject: [PATCH] WLANRTSYS-56718: IOP fixes implementation

- hostapd: wmm uapsd fix
  clear UAPSD flag in station info, if it unsupported by BSS
---
 src/ap/ieee802_11.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index cbb9e0f94..b00b0eaae 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3766,6 +3766,9 @@ static u16 check_wmm(struct hostapd_data *hapd, struct sta_info *sta,
 		sta->flags |= WLAN_STA_WMM;
 		wmm = (struct wmm_information_element *) wmm_ie;
 		sta->qosinfo = wmm->qos_info;
+		/* clear UAPSD if it unsupported by BSS */
+		if (!hapd->conf->wmm_uapsd)
+			sta->qosinfo &= ~WMM_QOSINFO_STA_AC_MASK;
 	}
 	return WLAN_STATUS_SUCCESS;
 }
-- 
2.43.0

